### PR TITLE
Fix Hibernate Envers native execution

### LINKED ISSUE #3120: Add native support for Hibernate Envers
**Description**
Hello,

is it possible to add Hibernate Envers module to easily audit JPA entities under Quarkus

Thanks by advance.

Damien

### PR DESCRIPTION
**Pull request by dcdh**
Fixes #3120

### CODE REVIEW

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmReflections.java [hunk 1]
```diff
@@ -102,6 +102,17 @@ public void registerCoreReflections(BuildProducer<ReflectiveClassBuildItem> refl
         simpleConstructor(reflectiveClass, "org.hibernate.internal.CoreMessageLogger_$logger");
     }
 
+    @BuildStep
+    public void registerEnversReflections(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        reflectiveClass.produce(
+                new ReflectiveClassBuildItem(false, false, "org.hibernate.tuple.component.DynamicMapComponentTuplizer"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, "org.hibernate.envers.DefaultRevisionEntity"));
+        reflectiveClass.produce(
+                new ReflectiveClassBuildItem(false, false, "org.hibernate.persister.entity.SingleTableEntityPersister"));
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmReflections.java]: This one is already registered above.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmReflections.java [hunk 2]
```diff
@@ -102,6 +102,17 @@ public void registerCoreReflections(BuildProducer<ReflectiveClassBuildItem> refl
...
+        reflectiveClass
+                .produce(new ReflectiveClassBuildItem(false, false, "org.hibernate.tuple.entity.DynamicMapEntityTuplizer"));
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmReflections.java]: Could you register the two with `org.hibernate.tuple` with the others above (at the beginning of the above method)?

We would only have the Envers-specific class here.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a few comments. There is something wrong with the tests, we have these failures:
```
2019-09-05 22:17:37,672 ERROR [org.hib.eng.jdb.spi.SqlExceptionHelper] (main) Connection is broken: "java.net.ConnectException: Connection refused: connect: localhost" [90067-197]
javax.persistence.PersistenceException: [PersistenceUnit: default] Unable to build Hibernate SessionFactory
```

I'll have to check CI.

**Comment by dcdh** [comment]
@gsmet I am aware about the build failing. Your CI is not involved it is my code. This PR is currently a draft one. I will ensure that the build is ok and made changes from all your comments to make this PR reviewable.

**Comment by dcdh** [comment]
To make the test run I add to add the same class `TestResources` like it was done in `integration-tests/hibernate-orm-panache`.

**Comment by dcdh** [comment]
I have got two remarks.

Can someone have a look on the CI because build fails on `Quarkus - Vert.x Web - Deployment` however I did not touch stuff related to Vert.x...

I have added a new project in `integration-tests` called `hibernate-orm-envers`. This project need to be added to the pipeline `Native Tests flyway, hibernate-orm-panache, reactive-pg-client` to be run as native application to ensure that hibernate envers is working in native mode. Can someone add it ?

**Comment by gsmet** [comment]
Ah. You’re having this issue too. I have it from time to time locally. 

@stuartwdouglas any idea what could cause this error that seems to be transient?

**Comment by gsmet** [comment]
I rebased and adjusted the CI pipeline. Let's see how it goes.

**Comment by gsmet** [comment]
Did you really get it to work in native? Because CI has this failure:
```
Caused by: org.hibernate.HibernateException: Generation of HibernateProxy instances at runtime is not allowed when the configured BytecodeProvider is 'none'; your model requires a more advanced BytecodeProvider to be enabled.
	at org.hibernate.bytecode.internal.none.DisallowedProxyFactory.getProxy(DisallowedProxyFactory.java:37)
	at org.hibernate.tuple.entity.AbstractEntityTuplizer.createProxy(AbstractEntityTuplizer.java:713)
	at org.hibernate.persister.entity.AbstractEntityPersister.createProxy(AbstractEntityPersister.java:4895)
	at org.hibernate.event.internal.DefaultLoadEventListener.createProxy(DefaultLoadEventListener.java:428)
	at org.hibernate.event.internal.DefaultLoadEventListener.createProxyIfNecessary(DefaultLoadEventListener.java:419)
	at org.hibernate.event.internal.DefaultLoadEventListener.proxyOrLoad(DefaultLoadEventListener.java:327)
	at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:108)
	at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:74)
	at org.hibernate.internal.SessionImpl.fireLoadNoChecks(SessionImpl.java:1298)
	at org.hibernate.internal.SessionImpl.internalLoad(SessionImpl.java:1161)
	at org.hibernate.type.EntityType.resolveIdentifier(EntityType.java:687)
	at org.hibernate.type.EntityType.resolve(EntityType.java:464)
	at org.hibernate.type.ManyToOneType.resolve(ManyToOneType.java:240)
	at org.hibernate.type.EntityType.resolve(EntityType.java:457)
	at org.hibernate.type.ComponentType.resolve(ComponentType.java:695)
	at org.hibernate.loader.Loader.extractKeysFromResultSet(Loader.java:878)
	at org.hibernate.loader.Loader.getRowFromResultSet(Loader.java:732)
	at org.hibernate.loader.Loader.processResultSet(Loader.java:1008)
	at org.hibernate.loader.Loader.doQuery(Loader.java:964)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:354)
	at org.hibernate.loader.Loader.doList(Loader.java:2810)
	at org.hibernate.loader.Loader.doList(Loader.java:2792)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2624)
	at org.hibernate.loader.Loader.list(Loader.java:2619)
	at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:506)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:396)
	at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:219)
	at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1547)
	at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1557)
	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1525)
```

We also have an issue with JDK 11 due to JAXB not being available in JDK 11 but this is easily fixable.

**Comment by cescoffier** [comment]
@dcdh can you rebase?

**Comment by gsmet** [comment]
I just rebased the PR. There's a good chance it won't pass CI anyway though as I didn't change the code.

@anamarija @irenakezic would one of you be interested in driving this puppy home?

**Comment by anamarija** [comment]
Ok. I will look into it later today.

**Comment by irenakezic** [comment]
Tried running the native test locally and I get the same error as stated above.

Caused by: javax.persistence.PersistenceException: org.hibernate.HibernateException: Generation of HibernateProxy instances at runtime is not allowed when the configured BytecodeProvider is 'none'; your model requires a more advanced BytecodeProvider to be enabled.

Will investigate this further tomorrow.

**Comment by gsmet** [comment]
In Quarkus, we entirely disable the bytecode provider because we do enhance the classes ourselves at build time.

If Envers needs to enhance classes somehow, we will need to do it at build time too and be sure Envers doesn't do it itself.

That being said, it might become a tricky issue then so I let you have a look but it might get quite complicated.

**Comment by gunnarmorling** [comment]
@gsmet et al., the Debezium community (personified in @Naros) is working on a Quarkus extension for supporting the outbox pattern. Our implementation makes use of Hibernate ORM's map entity mode in order to cater for a user-configurable structure of the outbox table.

To that end, we'd need two tuplizer implementations to be enabled for reflection, `DynamicMapEntityTuplizer` and `DynamicMapComponentTuplizer`, so that our extension can be used in native mode. I.e. not the full bytecode provider stuff discussed here.

Would you be willing to merge a PR to the ORM extension, if we provided such change? Thanks!

**Comment by gsmet** [comment]
I can extract this from Irena’s PR. I will create a PR this evening and ping you there.

**Comment by gunnarmorling** [comment]
Excellent, even better; Thanks a lot!

>

**Comment by gsmet** [comment]
So I extracted the reflection registration in another PR. The proxy issue will be solved as part of another issue. We should be in better shape after that.

**Comment by Sanne** [comment]
@gsmet , all, what is missing here?  I would really need the Envers support to be moved out of the ORM extension - I can work on creating a new Envers dedicated extension, but don't wish to step on other people's feet.

Can I help or would you rather prefer me waiting?

**Comment by gsmet** [comment]
@Sanne I haven't planned any work on this myself. I just extracted a small part of this PR to fix some issues for the Debezium team (merged a long time ago).

That being said, I think the IT from this PR and the small Envers reflection registration is a good base for starting again on this subject.

Your proxy code might have fixed one blocking issue. Not sure what's next.

And I agree having a separate extension is better.

**Comment by Sanne** [comment]
Many thanks for the summary @gsmet .  Anyone else was planning to take this one home? If not, I'm happy to take it.

**Comment by Sanne** [comment]
hi all, sorry for the delay: I'm finally back on this.

I will rebase this one, and incorporate it into a new dedicated extension for Envers.

**Comment by Sanne** [comment]
I'll close this as it's superseded by pr #10574 

Thanks @dcdh !

**Pull request closed without merging**